src/Razor/Microsoft.AspNetCore.Razor.Language/src/Components/ComponentTypeParamDirective.cs

```diff
@@ -7,23 +7,45 @@ namespace Microsoft.AspNetCore.Razor.Language.Components
 {
     internal class ComponentTypeParamDirective
     {
-        public static readonly DirectiveDescriptor Directive = DirectiveDescriptor.CreateDirective(
-            "typeparam",
-            DirectiveKind.SingleLine,
-            builder =>
-            {
-                builder.AddMemberToken(ComponentResources.TypeParamDirective_Token_Name, ComponentResources.TypeParamDirective_Token_Description);
-                builder.Usage = DirectiveUsage.FileScopedMultipleOccurring;
-                builder.Description = ComponentResources.TypeParamDirective_Description;
-            });
+        public static DirectiveDescriptor Directive = null;
 
-        public static RazorProjectEngineBuilder Register(RazorProjectEngineBuilder builder)
+        public static RazorProjectEngineBuilder Register(RazorProjectEngineBuilder builder, bool supportConstraints)
```

Comment:
Instead of passing this boolean through, should we just create a new `ComponentTypeParamWithGeneric` directive that we then conditionally register in the RazorEngineBuilder.

Reply:
I'm not sure if I'm understanding correctly.

The directives map 1 to 1 to language concepts, so it's not clear to me that is desirable to have two of them to represent the same concept. The problem with it is that we would need to litter the codebase with additional checks for the newly introduced directive, when the only difference is what is allowed per language version. 

I think it makes more sense for all that to be configured one time in one place since it only varies by language version.